File: app/src/main/java/com/weatherkremenchug/weatherkremenchugtest/activity/WeatherOnTodayActivity.java
package com.weatherkremenchug.weatherkremenchugtest.activity;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.RequiresApi;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Window;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.weatherkremenchug.weatherkremenchugtest.GPSTracker;
import com.weatherkremenchug.weatherkremenchugtest.R;
import com.weatherkremenchug.weatherkremenchugtest.data.WeatherOnToday;
import com.weatherkremenchug.weatherkremenchugtest.retrofit.CallToServer;

import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class WeatherOnTodayActivity extends AppCompatActivity {

    public boolean onSupportNavigateUp() {
        onBackPressed();
        finish();
        return true;
    }

    private static final String APP_ID = "c3eebf803f44713f50e31a7c5b215a73";
    String units = "metric";

    GPSTracker gpsTracker;

    double latitude;
    double longitude;

    ImageView image_weather_today;
    TextView textView_high_degree;
    TextView textView_low_degree;
    TextView textView_pressure;
    TextView textView_vlaznost;
    TextView textView_wind;
    TextView place;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_weather_on_today);
        Toolbar toolbar = findViewById(R.id.toolbar_weather_today);
        setSupportActionBar(toolbar);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            Objects.requireNonNull(getSupportActionBar()).setTitle("Погода на сегодня");
        }

        Window window = this.getWindow();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            window.setStatusBarColor(this.getResources().getColor(R.color.colorGray));
        }

        try {
            if (ContextCompat.checkSelfPermission(getApplicationContext(),
                    Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED ) {
                ActivityCompat.requestPermissions(this,
                        new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 101);
            }
        } catch (Exception e){
            e.printStackTrace();
        }

        gpsTracker = new GPSTracker(WeatherOnTodayActivity.this);

        if (gpsTracker.canGetLocation()) {
            latitude = gpsTracker.getLatitude();
            longitude = gpsTracker.getLongitude();
        } else {
            gpsTracker.showSettingsAlert();
        }

        image_weather_today = findViewById(R.id.image_weather_today);
        textView_high_degree = findViewById(R.id.degrees_high);
        textView_low_degree = findViewById(R.id.degrees_low);
        textView_pressure = findViewById(R.id.pressure_number);
        textView_vlaznost = findViewById(R.id.vlaga_number);
        textView_wind = findViewById(R.id.wind_number);
        place = findViewById(R.id.place);

        callWeather();

    }

    private void callWeather() {

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("http://api.openweathermap.org/data/2.5/")
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        CallToServer callToServer = retrofit.create(CallToServer.class);

        Call<WeatherOnToday> call = callToServer.getWeather(latitude, longitude, units, APP_ID);
        call.enqueue(new Callback<WeatherOnToday>() {
            @RequiresApi(api = Build.VERSION_CODES.KITKAT)
            @Override
            public void onResponse(@NonNull Call<WeatherOnToday> call, @NonNull Response<WeatherOnToday> response) {
                if (response.isSuccessful()) {
                    if (response.body() != null) {
                        textView_high_degree.setText(Objects.requireNonNull(response.body()).main.getTemp_max());
                        textView_low_degree.setText(Objects.requireNonNull(response.body()).main.getTemp_min());
                        textView_pressure.setText(Objects.requireNonNull(response.body()).main.getPressure());
                        textView_vlaznost.setText(Objects.requireNonNull(response.body()).main.getHumidity());
                        textView_wind.setText(Objects.requireNonNull(response.body()).wind.getSpeed());
                        place.setText(Objects.requireNonNull(response.body()).getName());

                        if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("04d")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_cloudy);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("02d")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_day_cloudy);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("01d")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_day_sunny);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("03d")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_cloud);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("09d")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_rain);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("10d")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_day_rain);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("11d")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_thunderstorm);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("13d")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_snow);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("50d")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_day_haze);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("01n")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_night_clear);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("02n")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_night_alt_partly_cloudy);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("03n")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_night_alt_cloudy);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("04n")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_cloud);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("09n")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_rain);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("10n")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_night_alt_rain);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("11n")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_thunderstorm);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("13n")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_snow);
                        } else if (Objects.requireNonNull(response.body()).weather.get(0).getIcon().equals("50d")) {
                            image_weather_today.setImageResource(R.drawable.ic_wi_night_fog);
                        }

                    }

                }
            }

            @Override
            public void onFailure(@NonNull Call<WeatherOnToday> call, @NonNull Throwable t) {
                Toast.makeText(WeatherOnTodayActivity.this, "Bad request", Toast.LENGTH_SHORT).show();
            }
        });

    }
}

File: app/src/main/java/com/weather/weathermain/activity/weathertoday/WeatherOnTodayActivity.kt
package com.weather.weathermain.activity.weathertoday

import android.Manifest
import android.annotation.SuppressLint
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.util.Log
import com.jakewharton.rxbinding2.view.RxView
import com.weather.weathermain.GPSTracker
import com.weather.weathermain.R
import com.weather.weathermain.activity.main.MainActivity
import com.weather.weathermain.activity.weathertoday.viewmodel.WeatherOnTodayViewModel
import com.weather.weathermain.data.WeatherOnTodayResponse
import com.weather.weathermain.utils.constants.APP_ID
import com.weather.weathermain.utils.constants.UNITS
import kotlinx.android.synthetic.main.activity_weather_on_today.*
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.time.LocalDate
import java.util.*

class WeatherOnTodayActivity : AppCompatActivity() {
    companion object {
        fun launch(context: Context) {
            val launcher = Intent(context, WeatherOnTodayActivity::class.java)
            context.startActivity(launcher)
        }
    }

    private var gpsTracker: GPSTracker? = null

    private var latitude: Double = 0.toDouble()
    private var longitude: Double = 0.toDouble()

    private lateinit var model: WeatherOnTodayViewModel

   private val calendar: Calendar = Calendar.getInstance()
    private val date: Date = calendar.time

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_weather_on_today)

        model = ViewModelProviders.of(this)[WeatherOnTodayViewModel::class.java]

        setupUi()

        try {
            if (ContextCompat.checkSelfPermission(applicationContext,
                            Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this,
                        arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), 101)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        gpsTracker = GPSTracker(this@WeatherOnTodayActivity)

        if (gpsTracker!!.canGetLocation()) {
            latitude = gpsTracker!!.latitude
            longitude = gpsTracker!!.longitude

        } else {
            gpsTracker!!.showSettingsAlert()
        }

        observeLiveData()
        setViewLiveData()
    }

    private fun observeLiveData() {
        model.requestCurrentWeather(latitude, longitude, UNITS, APP_ID)
    }

    private fun setViewLiveData() {
        model._getCurrentWeather().observe(this, Observer<WeatherOnTodayResponse> { data ->
            place.text = data!!.name
            if (data.weather[0].icon == "50d") iv_weather_today.setImageResource(R.drawable.icon_haze)
            tv_weather_name.text = data.weather[0].description
            tv_degree.text = data.main?.temp
            tv_humidity.text = data.main?.humidity
            tv_pressure.text = data.main?.pressure
            tv_wind_speed_value.text = data.wind?.speed

            tv_current_day.text = SimpleDateFormat("EEEE", Locale.getDefault()).format(date.time)
            tv_current_month.text = SimpleDateFormat("MMMM, d", Locale.getDefault()).format(date.time)

            tv_sunrise_value.text = SimpleDateFormat("hh:mm", Locale.getDefault()).format(data.sys?.sunrise!! * 1000L)
            tv_sunset_value.text = SimpleDateFormat("hh:mm", Locale.getDefault()).format(data.sys.sunset!! * 1000L)

        })

        model._getBackPressed().observe(this, Observer<Boolean> {
            if (it == true) {
                MainActivity.launch(this)
                finish()
            }
        })

        model._getDatFail().observe(this, Observer<String> { data_fail ->
            Log.d("DATA_IS_FAIL ", data_fail.toString())
        })

    }

    private fun setupUi() {
        setupClickListeners()
    }

    @SuppressLint("CheckResult")
    private fun setupClickListeners() {
//        RxView.clicks(btnBack)
//                .subscribe { model.onBackClicked() }
    }

//   when {
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "04d" -> image_weather_today?.setImageResource(R.drawable.ic_wi_cloudy)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "02d" -> image_weather_today?.setImageResource(R.drawable.ic_wi_day_cloudy)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "01d" -> image_weather_today?.setImageResource(R.drawable.ic_wi_day_sunny)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "03d" -> image_weather_today?.setImageResource(R.drawable.ic_wi_cloud)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "09d" -> image_weather_today?.setImageResource(R.drawable.ic_wi_rain)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "10d" -> image_weather_today?.setImageResource(R.drawable.ic_wi_day_rain)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "11d" -> image_weather_today?.setImageResource(R.drawable.ic_wi_thunderstorm)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "13d" -> image_weather_today?.setImageResource(R.drawable.ic_wi_snow)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "50d" -> image_weather_today?.setImageResource(R.drawable.ic_wi_day_haze)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "01n" -> image_weather_today?.setImageResource(R.drawable.ic_wi_night_clear)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "02n" -> image_weather_today?.setImageResource(R.drawable.ic_wi_night_alt_partly_cloudy)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "03n" -> image_weather_today?.setImageResource(R.drawable.ic_wi_night_alt_cloudy)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "04n" -> image_weather_today?.setImageResource(R.drawable.ic_wi_cloud)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "09n" -> image_weather_today?.setImageResource(R.drawable.ic_wi_rain)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "10n" -> image_weather_today?.setImageResource(R.drawable.ic_wi_night_alt_rain)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "11n" -> image_weather_today?.setImageResource(R.drawable.ic_wi_thunderstorm)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "13n" -> image_weather_today?.setImageResource(R.drawable.ic_wi_snow)
//  Objects.requireNonNull<WeatherOnTodayResponse>(response.body()).weather[0].icon == "50d" -> image_weather_today?.setImageResource(R.drawable.ic_wi_night_fog)
} // 178